test/server/config_validation/xds_fuzz.cc

```diff
@@ -0,0 +1,235 @@
+                                                 {buildCluster("cluster_0")}, {}, "1");
+  sendDiscoveryResponse<envoy::api::v2::ClusterLoadAssignment>(
+      Config::TypeUrl::get().ClusterLoadAssignment, {buildClusterLoadAssignment("cluster_0")},
+      {buildClusterLoadAssignment("cluster_0")}, {}, "1");
+
+  for (const auto& action : actions_) {
+    ENVOY_LOG_MISC(trace, "Action: {}", action.DebugString());
+    switch (action.action_selector_case()) {
+    case test::server::config_validation::Action::kAddListener: {
+      // Update the listener list.
+      listeners.erase(std::remove_if(listeners.begin(), listeners.end(),
+                                     [&action](const envoy::api::v2::Listener& listener) {
+                                       return listener.name() == action.add_listener().name();
+                                     }),
+                      listeners.end());
+      listeners.push_back(
+          buildListener(action.add_listener().name(), action.add_listener().route_config()));
+
+      // Send Listener DiscoveryResponse and update count.
+      updateListener(listeners, action.add_listener().version());
+      num_lds_updates_++;
+      test_server_->waitForCounterGe("listener_manager.lds.update_attempt", num_lds_updates_);
+      break;
+    }
+    case test::server::config_validation::Action::kAddRoute: {
+      // Update the route list.
+      std::remove_if(routes.begin(), routes.end(),
+                     [&](const envoy::api::v2::RouteConfiguration& route) {
+                       return route.name() == action.remove_route().name();
+                     });
+
+      routes.push_back(buildRouteConfig(action.add_route().name(), action.add_route().cluster()));
+      updateRoute(routes, action.add_route().version());
+      break;
+    }
+    case test::server::config_validation::Action::kRemoveListener: {
+      // Remove listener from list.
+      listeners.erase(std::remove_if(listeners.begin(), listeners.end(),
+                                     [&action](const envoy::api::v2::Listener& listener) {
+                                       return listener.name() == action.remove_listener().name();
+                                     }),
+                      listeners.end());
+      // Send state of the world listener DiscoveryResponse.
+      updateListener(listeners, action.add_listener().version());
+      num_lds_updates_++;
+      test_server_->waitForCounterGe("listener_manager.lds.update_attempt", num_lds_updates_);
+      break;
+    }
+    case test::server::config_validation::Action::kRemoveRoute: {
+      // Check if route with this name is in the list of routes, and if so, remove.
+      std::remove_if(routes.begin(), routes.end(),
+                     [&](const envoy::api::v2::RouteConfiguration& route) {
+                       return route.name() == action.remove_route().name();
+                     });
+      updateRoute(routes, action.add_route().version());
```

Comment:
So, by not sending the route, Envoy will actually just continue to latch the last route it received. This is because RDS/EDS are a bit weird for SoTW updates vs. LDS/CDS. The latter delete resources when not mentioned in an update, the former don't. It's weird, I know, and something I'm actively thinking about as part of the next generation of the UDPA transport protocol.